lib/lacerda/conversion/data_structure.rb

```diff
@@ -101,22 +100,62 @@ def add_properties_to_json_schema
         end
       end
 
-      def primitive_or_reference(type, is_required)
+      # returns the type of an array, given its specified type(s). This will be
+      # either exactly one basic type, or a oneOf in case there are 1+ types 
+      # or exactly 1 non-basic type. 
+      # As there are specied types in the array, `nil` should not be a valid value
+      # and therefore required should be true.
+      def array_items(types)
+        if types.size == 1 && PRIMITIVES.include?(types.first)
+          primitive(types.first, true) 
+        else
+          oneOf(types, true)
+        end
+      end
+
+      def primitive_or_oneOf(type, is_required)
       	return { 'type' => 'object' } if type.blank?
+        if PRIMITIVES.include?(type)
+          primitive(type, is_required)
+        else
+          oneOf([type], is_required)
+        end
+      end
 
+      # A basic type is either a primitive type with exactly 1 primitive type
+      #   {'type' => [boolean] }
+      # a reference
+      #   { '$ref' => "#/definitions/name" }
+      # or an object if the type in not there
+      #   { 'type' => object }
+      # Basic types don't care about being required or not.
+      def basic_type(type)
+        return { 'type' => 'object' } if type.blank?
         if PRIMITIVES.include?(type)
-          types = [type]
-          types << 'null' unless is_required
-          { 'type' => types }
+          primitive(type, false)
         else
-          types = [{'$ref' => "#/definitions/#{self.class.scope(@scope, type['literal'])}" }]
-          types << { 'type' => 'null' } unless is_required
-          {
-            'oneOf' => types
-          }
+          reference(type)
         end
       end
 
+      def oneOf(types, is_required)
```

Comment:
In case `is_required` is `false` and there is only 1 type (which happens in some places in the code), would it make more sense to have only a reference, instead of a `oneOf` of just one type?